components/brave_perf_predictor/browser/bandwidth_savings_predictor.cc

```diff
@@ -0,0 +1,151 @@
+        timing.paint_timing->first_contentful_paint.value().InMillisecondsF();
+
+  // Load
+  if (timing.document_timing->load_event_start.has_value())
+    feature_map_["metrics.observedLoad"] =
+        timing.document_timing->load_event_start.value().InMillisecondsF();
+
+  // Interactive
+  if (timing.interactive_timing->interactive.has_value())
+    feature_map_["metrics.interactive"] =
+        timing.interactive_timing->interactive.value().InMillisecondsF();
+}
+
+void BandwidthSavingsPredictor::OnSubresourceBlocked(
+    const std::string& resource_url) {
+  feature_map_["adblockRequests"] += 1;
+
+  ThirdPartyExtractor* extractor = ThirdPartyExtractor::GetInstance();
+
+  auto entity_name = extractor->GetEntity(resource_url);
+  if (entity_name.has_value())
+    feature_map_["thirdParties." + entity_name.value() + ".blocked"] = 1;
+}
+
+void BandwidthSavingsPredictor::OnResourceLoadComplete(
+    const GURL& main_frame_url,
+    const content::mojom::ResourceLoadInfo& resource_load_info) {
+  // If the resource load info comes without a valid corresponding
+  // main frame URL, ignore it
+  if (main_frame_url.is_empty() || !main_frame_url.has_host() ||
+      !main_frame_url.SchemeIsHTTPOrHTTPS()) {
+    return;
+  }
+  main_frame_url_ = main_frame_url;
+
+  bool is_third_party = !net::registry_controlled_domains::SameDomainOrHost(
+      main_frame_url, resource_load_info.url,
+      net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+
+  if (is_third_party) {
+    feature_map_["resources.third-party.requestCount"] += 1;
+    feature_map_["resources.third-party.size"] +=
+        resource_load_info.raw_body_bytes;
+  }
+
+  feature_map_["resources.total.requestCount"] += 1;
+  feature_map_["resources.total.size"] += resource_load_info.raw_body_bytes;
+  feature_map_["transfer.total.size"] +=
+      resource_load_info.total_received_bytes;
+  std::string resource_type;
+  switch (resource_load_info.resource_type) {
+    case content::ResourceType::kMainFrame:
+      resource_type = "document";
+      break;
+    case content::ResourceType::kSubFrame:
+      resource_type = "document";
+      break;
+    case content::ResourceType::kStylesheet:
+      resource_type = "stylesheet";
+      break;
+    case content::ResourceType::kScript:
+      resource_type = "script";
+      break;
+    case content::ResourceType::kImage:
+      resource_type = "image";
+      break;
+    case content::ResourceType::kFontResource:
+      resource_type = "font";
+      break;
+    case content::ResourceType::kMedia:
+      resource_type = "media";
+      break;
+    default:
+      resource_type = "other";
+      break;
+  }
+  feature_map_["resources." + resource_type + ".requestCount"] += 1;
+  feature_map_["resources." + resource_type + ".size"] +=
+      resource_load_info.raw_body_bytes;
+}
+
+double BandwidthSavingsPredictor::Predict() {
+  if (main_frame_url_.is_empty() || !main_frame_url_.has_host() ||
```

Comment:
mb `is_empty()` -> `is_valid()` ?

Reply:
why? it resets and terminates if the url is empty, not if it is valid

Reply:
sorry, i mean `main_frame_url_.is_empty()` -> `!main_frame_url_.is_valid()`